pkg/upgrade/control_plane.go

```diff
@@ -815,3 +1015,121 @@ func (u *ControlPlaneUpgrader) UpdateProviderIDsToNodes() error {
 
 	return nil
 }
+
+func (u *ControlPlaneUpgrader) waitForProviderID(ns, name string, timeout time.Duration) (string, error) {
+	u.log.Info("waitForMachineProviderID start", "namespace", ns, "name", name)
+	var providerID string
+	err := wait.PollImmediate(5*time.Second, timeout, func() (bool, error) {
+		machine := &clusterv1.Machine{}
+		if err := u.managementClusterClient.Get(context.TODO(), ctrlclient.ObjectKey{Name: name, Namespace: ns}, machine); err != nil {
+			return false, errors.WithStack(err)
+		}
+
+		if machine.Spec.ProviderID == nil {
+			return false, nil
+		}
+
+		providerID = *machine.Spec.ProviderID
+		if providerID != "" {
+			u.log.Info("Got providerID", "provider-id", providerID)
+			return true, nil
+		}
+		return false, nil
+	})
+
+	if err != nil {
+		return "", errors.Wrap(err, "timed out waiting for machine provider id")
+	}
+	return providerID, nil
+}
+
+func (u *ControlPlaneUpgrader) waitForMatchingNode(rawProviderID string, timeout time.Duration) (*v1.Node, error) {
+	u.log.Info("Waiting for node", "provider-id", rawProviderID)
+	var matchingNode v1.Node
+	providerID, err := noderefutil.NewProviderID(rawProviderID)
+	if err != nil {
+		return nil, err
+	}
+
+	err = wait.PollImmediate(5*time.Second, timeout, func() (bool, error) {
+		nodes, err := u.targetKubernetesClient.CoreV1().Nodes().List(metav1.ListOptions{})
+		if err != nil {
+			return false, err
```

Comment:
Same question here re `err` vs `timeout`

Reply:
again yes, logging the error is better than bailing early.

Reply:
I am ok with follow up issues too, we don't have to get everything into this PR. It's already pretty massive.